README
Map-Tube-Plugin-Formatter

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing,  you  can  find  documentation for this module with the perldoc
command.

    perldoc Map::Tube::Plugin::Formatter

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Map-Tube-Plugin-Formatter

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Map-Tube-Plugin-Formatter

    CPAN Ratings
        http://cpanratings.perl.org/d/Map-Tube-Plugin-Formatter

    Search CPAN
        http://search.cpan.org/dist/Map-Tube-Plugin-Formatter/

LICENSE AND COPYRIGHT

Copyright (C) 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Map::Tube::Plugin::Formatter

0.11  Wed Aug 31 10:50:00 2016
      - Added 'provides' to the Makefile.PL script.
      - Updated dependency on Map::Tube::Exception v3.04.

0.10  Wed Jul 13 10:05:00 2016
      - Updated pod with regard to format Map::Tube::Route.
      - Added SEE ALSO section to the pod.
      - Tidied up test scripts.

0.09  Wed May 04 20:35:00 2016
      - Added support for object Map::Tube::Route.
      - Updated copyright year information.

0.08  Sat Sep 19 04:45:00 2015
      - Replaced die() with exception Map::Tube::Exception::*.

0.07  Fri Mar 27 15:35:00 2015
      - Added code to the section "SYNOPSIS" to show formatting object of type Map::Tube::Line.

0.06  Tue Mar 24 10:50:00 2015
      - Added method to_string() to the package Map::Tube::Plugin::Formatter.
      - Tidied up pod document with regards to the above changes.

0.05  Mon Mar 23 18:10:00 2015
      - Updated email address for reporting bug in the pod document.

0.04  Mon Mar 23 11:40:00 2015
      - Tidied up pod document of the package Map::Tube::Plugin::Formatter.

0.03  Mon Mar 23 10:55:00 2015
      - Added new method get_data() to the package Map::Tube::Plugin::Formatter::Utils.
      - Refactored method to_json() and to_yaml() to make use of new method get_data().

0.02  Sun Mar 22 22:12:00 2015
      - Added support for YAML representation.

0.01  Sun Mar 22 11:20:00 2015
      - Module Created.


lib/Map/Tube/Plugin/Formatter.pm
package Map::Tube::Plugin::Formatter;

$Map::Tube::Plugin::Formatter::VERSION   = '0.11';
$Map::Tube::Plugin::Formatter::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Map::Tube::Plugin::Formatter - Formatter plugin for Map::Tube.

=head1 VERSION

Version 0.11

=cut

use 5.006;
use YAML;
use JSON qw();
use Map::Tube::Plugin::Formatter::Utils qw(xml get_data validate_object);

use Moo::Role;
use namespace::clean;

=head1 DESCRIPTION

A very simple add-on for L<Map::Tube> to format the supported objects.

=head1 SYNOPSIS

    use strict; use warnings;
    use Map::Tube::London;

    my $map = Map::Tube::London->new;

    my $node = $map->get_node_by_name('Baker Street');
    print $map->to_xml($node) ,   "\n\n";
    print $map->to_json($node),   "\n\n";
    print $map->to_yaml($node),   "\n\n";
    print $map->to_string($node), "\n\n";

    my $line = $map->get_line_by_name('Metropolitan');
    print $map->to_xml($line) ,   "\n\n";
    print $map->to_json($line),   "\n\n";
    print $map->to_yaml($line),   "\n\n";
    print $map->to_string($line), "\n\n";

    my $route = $map->get_shortest_route('Baker Street', 'Wembley Park');
    print $map->to_xml($route),   "\n\n";
    print $map->to_json($route),  "\n\n";
    print $map->to_yaml($route),  "\n\n";
    print $map->to_string($route),"\n\n";

=head1 SUPPORTED FORMATS

It currently supports the following formats.

=over 4

=item * XML

=item * JSON

=item * YAML

=item * STRING

=back

=head1 SUPPORTED OBJECTS

It currently supports the following objects.

=over 4

=item * L<Map::Tube::Node>

=item * L<Map::Tube::Line>

=item * L<Map::Tube::Route>

=back

=head1 METHODS

=head2 to_xml($object)

It takes an object (supported) and returns XML representation of the same.

=cut

sub to_xml {
    my ($self, $object) = @_;

    validate_object($object);

    my $data = {};
    if (ref($object) eq 'Map::Tube::Node') {
        $data = {
            node => {
                attributes => {
                    id     => $object->id,
                    name   => $object->name,
                },
                children   => {
                    link   => [ map {{ id => $_, name => $self->get_node_by_id($_)->name }} (split /\,/,$object->link) ],
                    line   => [ map {{ id => $_->id, name => $_->name                    }} (@{$object->line})         ],
                },
            },
        };
    }
    elsif (ref($object) eq 'Map::Tube::Line') {
        $data = {
            line => {
                attributes  => {
                    id      => $object->id,
                    name    => $object->name,
                    color   => $object->color || 'undef',
                },
                children    => {
                    station => [ map {{ id => $_->id, name => $_->name }} (@{$object->get_stations}) ],
                },
            },
        };
    }
    elsif (ref($object) eq 'Map::Tube::Route') {
        my $children = {};
        my $nodes    = $object->nodes;
        my $size     = $#$nodes;
        foreach my $i (1..($size-1)) {
            push @{$children->{node}}, { name => $nodes->[$i]->as_string, order => $i };
        }

        $data = {
            route => {
                attributes => {
                    from   => $object->from->as_string,
                    to     => $object->to->as_string,
                },
                children   => $children,
            },
        };
    }

    return xml($data);

}

=head2 to_json($object)

It takes an object (supported) and returns JSON representation of the same.

=cut

sub to_json {
    my ($self, $object) = @_;

    return JSON->new->utf8(1)->pretty->encode(get_data($self, $object));
}

=head2 to_yaml($object)

It takes an object (supported) and returns YAML representation of the same.

=cut

sub to_yaml {
    my ($self, $object) = @_;

    return Dump(get_data($self, $object));
}

=head2 to_string($object)

It takes an object (supported) and returns STRING representation of the same.

=cut

sub to_string {
    my ($self, $object) = @_;

    validate_object($object);

    return $object->as_string;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Map-Tube-Plugin-Formatter>

=head1 SEE ALSO

=over 4

=item * L<Map::Tube>

=back

=head1 BUGS

Please report any bugs or feature requests to C<bug-map-tube-plugin-formatter at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Map-Tube-Plugin-Formatter>.
I will  be notified and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Map::Tube::Plugin::Formatter

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Map-Tube-Plugin-Formatter>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Map-Tube-Plugin-Formatter>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Map-Tube-Plugin-Formatter>

=item * Search CPAN

L<http://search.cpan.org/dist/Map-Tube-Plugin-Formatter/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2015 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Map::Tube::Plugin::Formatter


lib/Map/Tube/Plugin/Formatter/Utils.pm
package Map::Tube::Plugin::Formatter::Utils;

$Map::Tube::Plugin::Formatter::Utils::VERSION   = '0.11';
$Map::Tube::Plugin::Formatter::Utils::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Map::Tube::Plugin::Formatter::Utils - Helper package for Map::Tube::Plugin::Formatter.

=head1 VERSION

Version 0.11

=cut

use vars qw(@ISA @EXPORT_OK);
require Exporter;
@ISA       = qw(Exporter);
@EXPORT_OK = qw(xml get_data validate_object);

use 5.006;
use strict; use warnings;
use Data::Dumper;
use XML::Twig;
use Map::Tube::Exception::MissingSupportedObject;
use Map::Tube::Exception::InvalidSupportedObject;
use Map::Tube::Exception::FoundUnsupportedObject;

=head1 DESCRIPTION

B<FOR INTERNAL USE ONLY>

=cut

sub xml {
    my ($data) = @_;

    my $twig = XML::Twig
        ->new()
        ->set_xml_version("1.0")
        ->set_encoding('utf-8');

    my $map = XML::Twig::Elt->new('map');
    $twig->set_root($map);

    foreach my $i (keys %$data) {
        my $e = $map->insert_new_elt($i, $data->{$i}->{attributes});
        foreach my $j (keys %{$data->{$i}->{children}}) {
            my $m = $e->insert_new_elt($j.'s');
            foreach (@{$data->{$i}->{children}->{$j}}) {
                $m->insert_new_elt($j, $_);
            }
        }
    }

    $twig->set_pretty_print('indented');
    return $twig->sprint;
}

sub get_data {
    my ($self, $object) = @_;

    validate_object($object);

    my $data = {};
    if (ref($object) eq 'Map::Tube::Node') {
        $data = {
            id    => $object->id,
            name  => $object->name,
            links => [ map {{ id => $_,     name => $self->get_node_by_id($_)->name }} (split /\,/,$object->link) ],
            lines => [ map {{ id => $_->id, name => $_->name                        }} (@{$object->line})         ],
        };
    }
    elsif (ref($object) eq 'Map::Tube::Line') {
        $data = {
            id       => $object->id,
            name     => $object->name,
            color    => $object->color || 'undef',
            stations => [ map {{ id => $_->id, name => $_->name }} (@{$object->get_stations}) ],
        };
    }
    elsif (ref($object) eq 'Map::Tube::Route') {
        my $children = [];
        my $nodes    = $object->nodes;
        my $size     = $#$nodes;
        foreach my $i (1..($size-1)) {
            push @{$children}, { name => $nodes->[$i]->as_string, order => $i };
        }

        $data = {
            from     => $object->from->as_string,
            to       => $object->to->as_string,
            nodes    => $children,
        };
    }

    return $data;
}

sub validate_object {
    my ($object) = @_;

    my @caller = caller(0);
    @caller = caller(2) if $caller[3] eq '(eval)';

    Map::Tube::Exception::MissingSupportedObject->throw({
        method      => __PACKAGE__."::validate_object",
        message     => "ERROR: No object received.",
        filename    => $caller[1],
        line_number => $caller[2] })
        unless defined $object;

    Map::Tube::Exception::InvalidSupportedObject->throw({
        method      => __PACKAGE__."::validate_object",
        message     => "ERROR: Invalid object received.",
        filename    => $caller[1],
        line_number => $caller[2] })
        unless ref($object);

    Map::Tube::Exception::FoundUnsupportedObject->throw({
        method      => __PACKAGE__."::validate_object",
        message     => "ERROR: Unsupported object received.",
        filename    => $caller[1],
        line_number => $caller[2] })
        unless ((ref($object) eq 'Map::Tube::Node')
                || (ref($object) eq 'Map::Tube::Line')
                || (ref($object) eq 'Map::Tube::Route'));
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Map-Tube-Plugin-Formatter>

=head1 BUGS

Please report any bugs or feature requests to C<bug-map-tube-plugin-formatter at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Map-Tube-Plugin-Formatter>.
I will  be notified and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Map::Tube::Plugin::Formatter::Utils

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Map-Tube-Plugin-Formatter>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Map-Tube-Plugin-Formatter>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Map-Tube-Plugin-Formatter>

=item * Search CPAN

L<http://search.cpan.org/dist/Map-Tube-Plugin-Formatter/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2015 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Map::Tube::Plugin::Formatter::Utils


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Map::Tube::Plugin::Formatter',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Map/Tube/Plugin/Formatter.pm',
    ABSTRACT_FROM      => 'lib/Map/Tube/Plugin/Formatter.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM                  => {
        'YAML'                 => '0.84',
        'JSON'                 => '2.53',
        'XML::Twig'            => '3.48',
        'Moo::Role'            => '2.000000',
        'namespace::clean'     => '0.24',
        'Map::Tube::Exception' => '3.04',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Map-Tube-Plugin-Formatter-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Map::Tube::Plugin::Formatter'        => { file => 'lib/Map/Tube/Plugin/Formatter.pm',       version => '0.11' },
            'Map::Tube::Plugin::Formatter::Utils' => { file => 'lib/Map/Tube/Plugin/Formatter/Utils.pm', version => '0.11' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Map-Tube-Plugin-Formatter.git',
                web  => 'https://github.com/manwar/Map-Tube-Plugin-Formatter',
            },
        }})
     : ()
    ),
);
